### PR TITLE
FindSFML: Missing dependency and wrong library path

### PR DESCRIPTION
**Pull request by Bromeon**
Two fixes for FindSFML.cmake:

1. Environment variables (e.g. PATH on Windows) are looked up early by CMake, and libraries found there are prioritized over those in SFML's own folder. To avoid paths to wrong libraries, this commit modifies the CMake `find_library()` call, such that environment variables are no longer considered as suitable paths.
See also: http://en.sfml-dev.org/forums/index.php?topic=17774

2. The dependency **libvorbisenc** is not listed in FindSFML. Thus, libraries using SFML get an incomplete `SFML_DEPENDENCIES` variable in CMake, leading to linker errors.

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
Could `NO_SYSTEM_ENVIRONMENT_PATH` have a negative effect on OS X ( @mantognini ) or Linux?

**Comment by mantognini** [comment]
I guess it should be fine. At least Jenkins didn't complain.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

**Comment by binary1248** [comment]
Isn't it possible to just add
```cmake
find_library(${output} NAMES ${ARGN} PATHS ${FIND_SFML_PATHS} PATH_SUFFIXES lib NO_SYSTEM_ENVIRONMENT_PATH)
```
before the original line without removing it? According to the [documentation](http://www.cmake.org/cmake/help/v3.0/command/find_library.html) calling find_library multiple times in succession is perfectly normal and is just a convoluted way of specifying a precedence order for which there is no mechanism by just calling the function a single time. This would also guarantee not breaking anything that currently works as well, because the original line remains as a fallback if the first call doesn't find anything.

**Comment by Bromeon** [comment]
That's what I suggested at the end [of this post](http://en.sfml-dev.org/forums/index.php?topic=17774.msg127728#msg127728). The question is whether it's necessary? Do we know systems that take library paths from environment variables?

In my opinion, we should avoid unneeded complexity in the configuration files... The CMake scripts are already a massive beast, at one point no one will understand them anymore :wink:

**Comment by eXpl0it3r** [comment]
> Do we know systems that take library paths from environment variables?

That's the question. Even if we don't know, it might still break somewhere. :wink:

**Comment by binary1248** [comment]
> The question is whether it's necessary? Do we know systems that take library paths from environment variables?

Since NO_SYSTEM_ENVIRONMENT_PATH only affects reading from PATH and LIB, we shouldn't generalize the question to all environment variables. The fact that we don't need to read from environment variables in the typical case doesn't mean we go breaking it for those hackers that do make use of this.

PATH isn't/shouldn't really be used to specify library search directories so I don't mind breaking this. _However_, we mustn't forget that even if we suppressed reading from PATH, during execution, the dynamic linker on Windows will try to resolve any imports using a certain directory search order. It begins in the current directory and at some point resorts to searching PATH. If we forbade CMake from searching PATH, we could link to a library that might not be in PATH although the PATH library would be the one that is loaded when the application is run. I wouldn't want to be the one who has to find out why applications start to mysteriously crash in this case. It works out most of the time, but I'm sure this was part of the rationale behind why CMake chooses to search PATH before anything else.

As for LIB, this is pretty much the standard environment variable where some Visual Studio users might append whatever library they intend on using. The idea is that you download some library with an installer, it installs its .dll files somewhere, adds that directory to PATH and modifies LIB to add the directory where the import libraries are located. This way, the installer can basically take care of everything for the developer automatically (isn't this the reason why people still use VS? :wink:). Excluding or de-prioritising LIB would break this for those people. They might have installed their own freetype or jpeg somewhere using some sort of installer, and intend to use that instead because it might have some new features. We're going to break this by de-prioritising LIB. MinGW went the other route and (AFAIK) hardcoded their search paths into the compiler/linker so they wouldn't have to make use of environment variables. This means that if you wanted a "fully automated" installer for some MinGW compatible library, it would have no choice than to dump them into the directories where the system headers/libraries are located. More messy, but no environment variables.

I remember the times when I still used my own version of GLEW (since it was newer than SFML's one) and it became really really annoying to always have to manually edit the CMake entries just because SFML insisted its own version is the one I want to use when building. If we can save less persistent users the time of having to do this, then I don't see why we shouldn't try to do so.

TL;DR: You are eventually going to break building for _someone_ by assuming reading from PATH and/or LIB is never necessary or even intentional. The original problem was that the search order was wrong _for us_. We solve that problem by searching _our own_ paths first, _then_ the environment paths if that failed. This might actually already be enough to "break" behaviour for some users on Windows, since they would be using a different jpeg in SFML and in their own code, so I don't know what to think of this. If you ask me, the "perfect" solution would be to search for .lib files first, no matter where they come from and then .dll files after that. The system directories should be preferred over our own if they are compatible (this can be/is tested in the Find scripts themselves if we specify a version).

> In my opinion, we should avoid unneeded complexity in the configuration files... The CMake scripts are already a massive beast, at one point no one will understand them anymore

This is more of a shortcoming of the CMake find_library command. If there was a way to specify the search order using a single command, we would probably do that, but there isn't. So, as stated in the documentation, we need to call it multiple times. I don't think applying something that is explicitly mentioned in the documentation constitutes making the scripts more complicated. If it ever gets improved one day, we could just combine them back into a single line again.

**Comment by Bromeon** [comment]
> We solve that problem by searching _our own_ paths first, _then_ the environment paths if that failed. 

Are LIB and PATH considered at all, even if we have a second `find_library` call? Since SFML ships the .lib files on Windows, they will always be found first -- unless the user manually removes them. So is the intended usage manual modification of SFML's installed lib directory?

> This might actually already be enough to "break" behaviour for some users on Windows, since they would be using a different jpeg in SFML and in their own code, so I don't know what to think of this. If you ask me, the "perfect" solution would be to search for .lib files first, no matter where they come from and then .dll files after that. The system directories should be preferred over our own if they are compatible (this can be/is tested in the Find scripts themselves if we specify a version).

Can we really assume that the user uses the system's libjpeg version? The user might as well use SFML's (not so unlikely if his whole game is written around SFML), or yet another one in his own local folder... So chances that we break something are there either way.

**Comment by binary1248** [comment]
Well, I guess since this only really affects the building of SFML, I guess we can live with that. If we ship every single dependency ourselves anyway and only rely on OpenGL and the system libraries, I don't think anything should go wrong even if we excluded more stuff from the search paths.

**Comment by mantognini** [comment]
> If we ship every single dependency ourselves anyway and only rely on OpenGL and the system libraries

That's not 100% true: when building SFML for package managers one should probably use the dependencies from that package manager. That's how homebrew work at least. (I guess it's always possible to use cmake variable to specify where deps should be taken from but that's not really convenient.)

**Comment by binary1248** [comment]
That's the question. The way I see it `NO_SYSTEM_ENVIRONMENT_PATH` essentially means "don't let the user use their own installed libraries _on Windows_". I also find it misleading because other environment variables are still read and even used on other operating systems. This is just a Windows thing, and because Windows doesn't have a package manager it becomes more tedious for developers who know what they are doing and have to go and manually alter CMake entries at the end of the day.

As usual, it's all Windows' fault :grin:.

**Comment by eXpl0it3r** [comment]
So what's the verdict?

**Comment by binary1248** [comment]
I guess if it doesn't change anything for users who intentionally override SFML's supplied extlibs with their own, it shouldn't make a difference. People who count on PATH and LIB to override any of SFML's extlibs are doing something wrong to begin with anyway :grin:.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.
